docs/madr/decisions/014-mesh-circuit-breaker-policy.md

```diff
@@ -0,0 +1,198 @@
+
+- flattening `detectors.*` options with only single (`.consecutive`) property
+- changing word `Errors` to `Failures`
+- adding `origin` suffix to the local failures detector
+
+It means `totalErrors.consecutive`, `gatewayErrors.consecutive` and `localErrors.consecutive`
+becomes: `consecutiveTotalFailures`, `consecutiveGatewayFailures` and `consecutiveLocalOriginFailures` respectively.
+
+### Current configuration
+
+Below is the sample `MeshCircuitBreaker` configuration
+
+```yaml
+conf:
+  interval: 1s
+  baseEjectionTime: 30s
+  maxEjectionPercent: 20
+  splitExternalAndLocalErrors: false
+  thresholds:
+    maxConnections: 2
+    maxPendingRequests: 2
+    maxRequests: 2
+    maxRetries: 2
+  detectors:
+    totalErrors:
+      consecutive: 20
+    gatewayErrors:
+      consecutive: 10
+    localErrors:
+      consecutive: 7
+    standardDeviation:
+      requestVolume: 10
+      minimumHosts: 5
+      factor: 1.9
+    failure:
+      requestVolume: 10
+      minimumHosts: 5
+      threshold: 85
+
+```
+
+### Specification
+
+#### Top level
+
+As a part of this MADR root `targetRef`'s kind can be configured as one
+of `Mesh|MeshSubset|MeshService|MeshServiceSubset`. When we finally introduce `MeshHTTPRoute` kind and if there will be
+a reasonable use case, in the next iteration `MeshGatewayRoute` and `MeshHTTPRoute` could be added as a viable kinds.
+
+```yaml
+targetRef:
+  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+  name: ...
+```
+
+#### From level
+
+`MeshCircuitBreaker` is outbound only policy, so only `to` should be configured.
+
+#### To level
+
+```yaml
+to:
+- targetRef:
+    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+    name: ...
+```
+
+#### Minimal configuration
+
+As all the connection pool related properties have default values, there is no need to specify any of these to apply the
+policy
+
+```yaml
+default:
+  connectionPool: { }
+```
+
+### Examples
+
+#### Default configuration
+
+##### `MeshCircuitBreaker`
+
+```yaml
+type: MeshCircuitBreaker
+mesh: default
+name: default-circuit-breaker
+spec:
+  targetRef:
+    kind: Mesh
+    name: default
+  to:
+  - targetRef:
+      kind: Mesh
+      name: default
+    default:
+      connectionPool:
+        maxConnections: 1024
+        maxPendingRequests: 1024
+        maxRetries: 3
+        maxRequests: 1024
+```
+
+#### Extensive configuration
+
+```yaml
+type: MeshCircuitBreaker
+mesh: default
+name: extensive-circuit-breaker
+spec:
+  targetRef:
+    kind: Mesh
+    name: default
+  to:
+  - targetRef:
+      kind: Mesh
+      name: default
+    default:
+      connectionPool:
+        maxConnections: 1024
+        maxPendingRequests: 1024
+        maxRetries: 3
+        maxRequests: 1024
+      outlierDetection:
+        interval: 5s
+        baseEjectionTime: 30s
+        maxEjectionPercent: 20
+        splitExternalAndLocalErrors: true
+        detectors:
+          consecutiveTotalFailures: 10
```

Comment:
totalErrors is weird as in envoy it's total 5xx errors which I believe is a lot more explicit

Reply:
I would agree if not the case when we are talking about HTTP only, but for TCP there is no 5xx and I think it's then misleading. I don't have a strong opinion though

Reply:
What I meant is that this always configures 5xx:

https://github.com/kumahq/kuma/blob/cddddc99fca74a8eccf0ad1719de48a67fd33c6a/pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go#L35-L42

Here by not using 5xx in the name we hide the fact that this will only work wiht HTTP

Reply:
but according to envoy docs it's not the case:
> consecutive_5xx
>    ([UInt32Value](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#uint32value)) The number of consecutive server-side error responses (for HTTP traffic, 5xx responses; for TCP traffic, connection failures; for Redis, failure to respond PONG; etc.) before a consecutive 5xx ejection occurs. Defaults to 5.
> ref: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/outlier_detection.proto#envoy-v3-api-msg-config-cluster-v3-outlierdetection

Reply:
Ok my bad it's the envoy field that is missnamed :)